File: LeetCode/cpp/307.cpp
class NumArray {
    vector<int> array;
public:
    NumArray(vector<int> nums) {
        int n = nums.size();
        array = vector<int>(2 * n, 0);
        for (int i = n, j = 0; i < 2 * n; ++i, ++j)
            array[i] = nums[j];
        for (int i = n - 1; i > 0; --i)
            array[i] = array[i * 2] + array[i * 2 + 1];
    }

    void update(int i, int val) {
        i += array.size() / 2;
        int prev = array[i];
        while (i > 0) {
            array[i] += val - prev;
            i /= 2;
        }
    }

    int sumRange(int i, int j) {
        int sum = 0;
        i += array.size() / 2;
        j += array.size() / 2;
        while (i <= j) {
            if (i % 2 == 1) {
                sum += array[i];
                ++i;
            }
            if (j % 2 == 0) {
                sum += array[j];
                --j;
            }
            i /= 2;
            j /= 2;
        }
        return sum;
    }
};

File: LeetCode/python3/307.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.vector = [0 for _ in range(2 * self.n)]
        j = 0
        for i in range(self.n, 2 * self.n):
            self.vector[i] = nums[j]
            j += 1
        for i in range(self.n - 1, 0, -1):
            self.vector[i] = self.vector[i * 2] + self.vector[i * 2 + 1]

    def update(self, i: int, val: int) -> None:
        i += self.n
        prev = self.vector[i]
        while i > 0:
            self.vector[i] += val - prev
            i //= 2

    def sumRange(self, i: int, j: int) -> int:
        sum = 0
        i += self.n
        j += self.n
        while i <= j:
            if i % 2 == 1:
                sum += self.vector[i]
                i += 1
            if j % 2 == 0:
                sum += self.vector[j]
                j -= 1
            i //= 2
            j //= 2
        return sum
